// @nomic-ai/atlas/dist/organization.js
import { get_env_user } from './user.js';
import { AtlasProject } from './project.js';
export class AtlasOrganization {
    constructor(id, user) {
        this._info = undefined;
        this.id = id;
        this.user = user || get_env_user();
    }
    info() {
        if (this._info !== undefined) {
            return this._info;
        }
        this._info = this.user.apiCall(`/v1/organization/${this.id}`, 'GET');
        return this._info;
    }
    async projects() {
        const info = (await this.info());
        return info.projects;
    }
    async create_project(options) {
        const info = (await this.info());
        const user = this.user;
        if (options.unique_id_field === undefined) {
            throw new Error('unique_id_field is required');
        }
        if (options.project_name === undefined) {
            throw new Error('Project name is required');
        }
        options['modality'] = options['modality'] || 'text';
        const data = (await user.apiCall(`/v1/project/create`, 'POST', {
            ...options,
            organization_id: this.id,
        }));
        return new AtlasProject(data['project_id'], user);
    }
}
